### PR TITLE
Tpwdcee 6463 feature

### PR DESCRIPTION
**Pull request by edra-bumci**

### CODE REVIEW

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/classes/Notification/Initial/Success.php  1
- wirecardpaymentgateway/classes/Notification/Failure.php  1
- wirecardpaymentgateway/classes/Notification/PostProcessing/Success.php  3
- wirecardpaymentgateway/classes/Response/PostProcessing/Success.php  3
- wirecardpaymentgateway/classes/Service/OrderStateNumericalValues.php  1
- wirecardpaymentgateway/helper/Service/OrderService.php  2
- wirecardpaymentgateway/classes/Response/Failure.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=5492081&bid=17998485)

File: wirecardpaymentgateway/helper/Service/OrderService.php [hunk 1]
```diff
@@ -31,36 +35,46 @@ public function __construct($order)
...
+            case TransactionTypes::TYPE_REFUND_CAPTURE:
+            case TransactionTypes::TYPE_REFUND_DEBIT:
+            case TransactionTypes::TYPE_VOID_PURCHASE:
+                return true;
+            default:
+                return false;
+        }
+    }
+
+	/**
```
Comment (rinnhofer) [inline wirecardpaymentgateway/helper/Service/OrderService.php:72]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=5492081)

File: wirecardpaymentgateway/helper/Service/OrderService.php [hunk 2]
```diff
@@ -31,36 +35,46 @@ public function __construct($order)
...
+	 * @param string $transaction_id
+	 *
```
Comment (rinnhofer) [inline wirecardpaymentgateway/helper/Service/OrderService.php:74]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=5492081)

File: wirecardpaymentgateway/helper/Service/OrderService.php [hunk 3]
```diff
@@ -31,36 +35,46 @@ public function __construct($order)
...
+	 * @since 2.1.0
```
Comment (rinnhofer) [inline wirecardpaymentgateway/helper/Service/OrderService.php:75]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=5492081)

File: wirecardpaymentgateway/helper/Service/OrderService.php [hunk 4]
```diff
@@ -31,36 +35,46 @@ public function __construct($order)
...
+            case TransactionTypes::TYPE_REFUND_DEBIT:
+            case TransactionTypes::TYPE_VOID_PURCHASE:
+                return true;
+            default:
+                return false;
+        }
+    }
+
+	/**
+	 * @param string $transaction_id
```
Comment (rinnhofer) [inline wirecardpaymentgateway/helper/Service/OrderService.php:73]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=5492081)

File: wirecardpaymentgateway/helper/Service/OrderService.php [hunk 5]
```diff
@@ -31,36 +35,46 @@ public function __construct($order)
...
+	 *
+	 * @since 2.1.0
+	 */
```
Comment (rinnhofer) [inline wirecardpaymentgateway/helper/Service/OrderService.php:76]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=5492081)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- wirecardpaymentgateway/classes/Response/PostProcessing/Success.php  3
- wirecardpaymentgateway/classes/Service/OrderStateNumericalValues.php  1
- wirecardpaymentgateway/classes/Notification/Failure.php  1
- wirecardpaymentgateway/classes/Notification/Initial/Success.php  1
- wirecardpaymentgateway/classes/Notification/PostProcessing/Success.php  3
- wirecardpaymentgateway/helper/Service/OrderService.php  2
- wirecardpaymentgateway/classes/Response/Failure.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=5492081&bid=17998485)

File: wirecardpaymentgateway/helper/Service/OrderService.php [hunk 6]
```diff
@@ -31,36 +35,46 @@ public function __construct($order)
...
-     * @param string $transaction_id
-     * @param float $amount
-     * @since 2.1.0
+     * @param Transaction $transaction
+     *
+     * @return bool
+     * @since 2.10.0
      */
-    public function updateOrderPayment($transaction_id, $amount)
+    public function createOrderPayment($transaction)
```
Comment (tomazpu) [inline wirecardpaymentgateway/helper/Service/OrderService.php:43]: In general, I think coupling it with a transaction type is wrong. As in the acceptance criteria, the payments are coupled to order states I would make the check to an order state and not the transaction type

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/classes/Response/Initial/Success.php  -1
+ wirecardpaymentgateway/helper/Service/OrderService.php  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=5492081&bid=17998485)

File: wirecardpaymentgateway/helper/OrderManager.php [hunk 7]
```diff
@@ -63,6 +67,7 @@ class OrderManager
     public function __construct()
     {
         $this->module = \Module::getInstanceByName(\WirecardPaymentGateway::NAME);
+        $this->order_service = new OrderService(null);
```
Comment (tomazpu) [inline wirecardpaymentgateway/helper/OrderManager.php:70]: try to get the order from order reference

File: wirecardpaymentgateway/helper/Service/OrderService.php [hunk 8]
```diff
@@ -9,6 +9,8 @@
 
 namespace WirecardEE\Prestashop\Helper\Service;
 
+use Db;
```
Comment (tomazpu) [inline wirecardpaymentgateway/helper/Service/OrderService.php:12]: \Db

File: wirecardpaymentgateway/helper/Service/OrderService.php [hunk 9]
```diff
@@ -31,36 +33,42 @@ public function __construct($order)
...
-
-        $last_index = count($order_payments) - 1;
-
-        if (!empty($order_payments)) {
-            $order_payments[$last_index]->transaction_id = $transaction_id;
-            $order_payments[$last_index]->amount = $amount;
-            $order_payments[$last_index]->save();
+        $orderState = $this->order->current_state;
+        if ($this->isOrderPaymentCreate($orderState)) {
+            $amount = -1 * $this->order->total_paid;
```
Comment (tomazpu) [inline wirecardpaymentgateway/helper/Service/OrderService.php:45]: cast to float total_paid

File: wirecardpaymentgateway/helper/Service/OrderService.php [hunk 10]
```diff
@@ -31,36 +33,42 @@ public function __construct($order)
...
+    /**
+     * @param string $orderState
+     *
+     * @return bool
+     * @since 2.10.0
+     */
+    public function isOrderPaymentCreate($orderState)
+    {
+        switch ($orderState) {
+            case '7':
```
Comment (tomazpu) [inline wirecardpaymentgateway/helper/Service/OrderService.php:59]: \Configuration::get('PS_OS_REFUND')

lookup OrderStateMappingDefinition.php

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/helper/Service/OrderService.php  -1
+ wirecardpaymentgateway/classes/Response/Initial/Success.php  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=5492081&bid=17998485)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/helper/Service/OrderService.php  -1
+ wirecardpaymentgateway/classes/Response/Initial/Success.php  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=5492081&bid=17998485)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/helper/Service/OrderService.php  -1
+ wirecardpaymentgateway/classes/Response/Initial/Success.php  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=5492081&bid=17998485)

File: wirecardpaymentgateway/helper/OrderManager.php [hunk 11]
```diff
@@ -90,13 +100,19 @@ public function createOrder($cart, $state, $paymentType)
             $cart->secure_key
         );
 
+        $orderReference = $this->module->currentOrderReference;
```
Comment (tomazpu) [inline wirecardpaymentgateway/helper/OrderManager.php:103]: you can put it before line 89 and use the variable also there :)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage increased per file
===========================
+ wirecardpaymentgateway/classes/Response/Initial/Success.php  5
+ wirecardpaymentgateway/models/Transaction.php  1
         

Coverage decreased per file
===========================
- wirecardpaymentgateway/helper/Service/OrderService.php  -1
- wirecardpaymentgateway/helper/OrderManager.php  -3
         

Complexity decreasing per file
==============================
+ wirecardpaymentgateway/classes/Response/Initial/Success.php  -1
+ wirecardpaymentgateway/helper/Service/OrderService.php  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=5492081&bid=17998485)

File: wirecardpaymentgateway/helper/Service/OrderService.php [hunk 12]
```diff
@@ -31,59 +33,72 @@ public function __construct($order)
...
+    public function addTransactionIdToOrderPayment($transaction_id)
     {
         $order_payments = \OrderPayment::getByOrderReference($this->order->reference);
-
         $last_index = count($order_payments) - 1;
+        $order_current_state = (int) $this->order->getCurrentState();
+        $order_payment_state = (int) \Configuration::get('PS_OS_PAYMENT');
 
-        if (!empty($order_payments)) {
+        if (!empty($order_payments)&&($order_current_state === $order_payment_state)) {
```
Comment (tomazpu) [inline wirecardpaymentgateway/helper/Service/OrderService.php:78]: white spaces missing between the two conditions :), also I suggest making an array check as we will add partly paid but if you think you do in the next US its ok with me. Then you can skip this.

Comment (edra-bumci) [reply wirecardpaymentgateway/helper/Service/OrderService.php:78]: I would say it is better to do it in the next US, as I dont really have the exact image of what I have to check there in my mind

**Pull request merged**